app/webpacker/controllers/bulk_form_controller.js

```diff
@@ -161,4 +163,34 @@ export default class BulkFormController extends Controller {
       return element.defaultValue !== undefined && element.value != element.defaultValue;
     }
   }
+
+  #removeAnimationClasses(productRowElement) {
+    productRowElement.classList.remove('slide-in');
+    productRowElement.removeEventListener('animationend', this.#removeAnimationClasses.bind(this, productRowElement));
+  }
+
+  #observeProductsTableRows() {
+    this.productsTableObserver = new MutationObserver((mutationList, _observer) => {
+      const mutationRecord = mutationList[0];
+
+      if(mutationRecord) {
+        // Right now we are only using it for product clone, so it's always first
+        const productRowElement = mutationRecord.addedNodes[0];
+
+        if (productRowElement) {
+          productRowElement.addEventListener('animationend', this.#removeAnimationClasses.bind(this, productRowElement));
+          // This is equivalent to form.elements.
+          const productRowFormElements = productRowElement.querySelectorAll('input, select, textarea, button');
+          this.#registerElements(productRowFormElements);
+          this.toggleFormChanged();
+        }
+      }
+    });
+
+    const productsTable = document.querySelector('.products');
+    // Above mutation function will trigger,
+    // whenever +products+ table rows (first level children) are mutated i.e. added or removed
+    // right now we are using this for product clone 
+    this.productsTableObserver.observe(productsTable, { childList: true });
+  }
```

Comment:
Nice work, I like that the code is building upon basic knowledge of the DOM (rows can be added/removed, and we need to handle them somehow. The `.products` selector could probably be simply `table`, which would make it not product-specific.

However the BulkFormController, is getting quite large, and the new feature doesn't seem core to the goal of:
```
// Manage "changed" state for a form with multiple records
```

So was thinking it would be great to split it into a separate controller for handling this. We could use  a Stimulus Outlet to trigger the registerElement action (or there might be a better way). I think it might almost be as simple as:

clone.turbo_stream:
```html
%tbody{ data-controller="cloned-resource" data-cloned-resource-outlet="products-form" }
```

```javascript
class ClonedResource extends Controller {
connect() {
   bulkFormOutlet.registerElements();

  this.addEventListener('animationend', this.#removeAnimationClasses.bind(this)); // assuming controller is initialised before animation finishes. But that's probably a safe assumption
}
...
```

Smaller modules will probably help with maintenance and re-use in the future, but maybe it risks creating a hug sea of controllers that is hard to navigate.. 

Would you be willing to give it a try?

Reply:
Sigh, putting on a project manager hat 🤠 , I realise that we need more focus on just getting this released. I think the current solution is fine, so I would suggest we can merge this. 
And if you would like to, we can deal with this as a tech debt issue later.

Reply:
That's a great idea. Sure, I'd love to work on this in a tech debt. It would be fun 😀

Reply:
Your suggestion is way better than me using a mutation observer😂. I just realized.